### ISSUE TITLE
Minimalist Positional/Hierarchical Audio UI

### ISSUE THREAD

**Issue by FelixBenning** [mumble-voip/mumble#4273]
**Context**
Due to the Coronavirus University Lectures moved online, in online lectures people tend to talk less with each other. Small personal chit chat does not really happen as much. It is difficult to break up a group into smaller groups on the spot (although Zoom provides a little used breakout rooms feature). More specifically on the Academia Stack Exchange someone argued, that people at conferences would form small groups and go to lunch together in person, talking about their current research interests. This type of talk does not really happen in online meetings, which is a great loss for brainstorming ideas.

What makes people loose up in person is the ability to target what they say at people in proximity or at the entire room. Multiple conversations can happen in parallel since distance quiets down speakers from other groups. You can easily drop in and out of conversations by moving around the room.

Talking about this issue with a friend I suggested that VR is going to fix that issue at some point. Later that day I realized that VR is not really necessary to fix this. What is necessary is a virtual space, not necessarily *being* in that virtual space. That is when I remembered a video where people used mumble for positional audio in a game. So in principle you could just hold your conference in minecraft and get the desired properties. But asking people to join a minecraft server for a conference is going to sound strange. And you are not really interested in any of the games features. The only thing you really want is the coordinates from it. 

**Describe the feature you have in mind**
Create an alternative UI to the current list of people in a Room. This UI would be
 - an empty two dimensional plane people (avatars, i.e. profile picture bubbles) are placed into
 - people can move around by clicking on a spot (teleporting) or moving with the arrow keys
 - there would be two different audio keys:
   - a positional audio key: If you talk, people in your proximity hear what you say and the sound drops off
   - an address-the-room key: if you talk, everyone can hear you in full volume
 - You can draw rectangles into the empty plane, this creates virtual rooms. 
   - Audio using the positional key immediately drops off outside the room
   - you can select a room by clicking on a rooms walls with your mouse, highlighting it with a different colour, the address-the-room key speaks in full volume to the entire room you selected,  
   - clicking outside every room unselects all the rooms and "address-the-room" speaks in full volume to the entire floor
   - rooms inside of rooms are of course possible too

You already have:
 - positional audio
 - channels and subchannels (rooms)

Needs to be added:
 - Open an empty window
 - Render profile pictures inside this window
 - allow for manipulation of your own coordinates
 - render rectangles and place people in channels/subchannels depending on their location and highlighting of rectangles

I have never done any UI work, but this is probably quite straightforward. Collision checks are not even desired, checking whether someone is inside a rectangle only requires checking if the x coordinates is in a certain interval/ y is in a certain interval. So I do not think this should be too difficult. Depending how entangled the code is, the rest should be easy to add as well. (I heard you are planning a refactor?)

Since the refactor plans to separate backend and frontend, this would simply be a different frontend. Although it might still be nice, to have a list of participants etc. So I guess this would only be added as a front-end module. 

This could be a killer feature as no other conferencing tool has positional audio - according to my google foo. This could add some publicity, which might help with the refactor.

**Comment by Krzmbrzl** [comment]
I guess the Manual plugin is kinda what you are looking for (of the very basic feature). That allows you to position yourself in a 3D room (influencing the audio).

I'm not sure though whether it shows the position of others in the room :thinking: 

The functionality you described could definitely built on top of the Manual plugin.

> I heard you are planning a refactor?

yes

> Since the refactor plans to separate backend and frontend, this would simply be a different frontend. Although it might still be nice, to have a list of participants etc. So I guess this would only be added as a front-end module.

I think a module might be better.

For the positionaing of entire channels in 2D space, you'd need additional info that needs to be shared by the clients though. This is currently not possible with the current protocol. I also don't really think this is really necessary. Imo the 2D plane should simply represent a single room in which you can position yourself as you wish. If you want to hop into another channel, just use the regular UI (same for creating a new one) :thinking:

**Comment by FelixBenning** [comment]
I had a look at the plugin and this seems indeed like the thing I am looking for. I'll probably have a deeper look once I have a bit more time in a month or so. Maybe I can actually do that myself. The only thing I am worried about is exchanging the coordinates between clients. I do not really want to get into networking and crypto, no idea how that works - any chance you could piggyback that on the existing mumble server?

Though I guess that is something which I could deal with once I have a running demo. Qt tutorial for future me: https://www.youtube.com/watch?v=8ntEQpg7gck

> I think a module might be better.
you mean plugin?

>For the positioning of entire channels in 2D space, you'd need additional info that needs to be shared by the clients though. This is currently not possible with the current protocol. I also don't really think this is really necessary. Imo the 2D plane should simply represent a single room in which you can position yourself as you wish. If you want to hop into another channel, just use the regular UI
The proximity based based thing is supposed to get people talking. Once they are talking and the soundstage sounds like a crowded cafeteria, you would want to introduce some sound barriers. Of course you could now create a different channel, interrupt the flow of conversation and ask people to join that other channel or find all the names in a list having nothing to do with your proximity based sense of a group and move them yourself. Then you will have to wait a few seconds for everyone to find where the channel is and how to move themselves there making sure that everyone is there. Since you now entered a different channel all the proximities get reshuffled as well, since your channels would be the "floors". This feels a lot more disruptive than simply drawing a rectangle around an existing group on the fly. 

That is the reason why I think such a thing would be cool. But it is also definitely something which can be added after you get a minimum working example.

**Comment by Krzmbrzl** [comment]
> A feature with even lower priority would be screen-sharing (presentations popping up) and video, which links into #4150 I guess

That's completely unrelated to this feature request though.

> you mean plugin?

No I meant  module but now that I re-read your original question I realized that my answer doesn't make a whole lot of sense xD
Take as an agreement to your suggestion.

> any chance you could piggyback that on the existing mumble server?

Sure - you can simply create new ProtoBuf messages for transmitting your data and let the server relay them to the respective clients.

> This feels a lot more disruptive than simply drawing a rectangle around an existing group on the fly.

But it is the only way you can use Mumble's positional audio feature as-is. For your suggestion you'd have to rewrite/extend it.

**Comment by Krzmbrzl** [comment]
As I don't think this is a feature that the Mumble client will implement due to it being a rather niche-approach to UI, I'll close this issue...
